### PR TITLE
Refactor Specifications/Serializers

### PR DESCRIPTION
**Pull request by vepkenez**
* refactors code around input and output specifications and serializers
* removes 100+ lines of controllers/serializers code
* adds new declarative format for io specification using py-marshmallow
* adds improved validation for different datatypes
* additional extensibility to validate keys, treasure maps, message kits etc.
* allows for specifications to indicate type which enables better connect native integration as well as improved compatibility with REST/OPTIONS interfaces
* required by #1299

### CODE REVIEW

File: nucypher/characters/control/specifications/fields/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from marshmallow import fields
+from .key import *
+from .treasuremap import *
+from .messagekit import *
+from .datetime import *
+from .label import *
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/fields/__init__.py]: Should we just add all field type definitions to the same `fields.py` file instead of one per file? Would that save us from needing this code?

File: nucypher/characters/control/specifications/fields/messagekit.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+class MessageKitField(fields.Field):
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return b64encode(bytes(value)).decode()
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        if isinstance(value, bytes):
+            return value
+        return b64decode(value)
+
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/fields/messagekit.py]: How come `MessageKit` isn't validated? i.e. `def _validate(...)`

File: nucypher/characters/control/specifications/fields/treasuremap.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+from base64 import b64decode, b64encode
+
+class TreasureMapField(fields.Field):
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return b64encode(bytes(value)).decode()
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        return b64decode(value)
+
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/fields/treasuremap.py]: Same question as `MessageKit`: 

How come TreasureMap isn't validated? i.e. def _validate(...)

File: nucypher/characters/control/specifications/enrico.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from .fields import fields
+from .base import BaseSchema
+
+class EncryptMessage(BaseSchema):
+
+    message = fields.Str(required=True, load_only=True)
+    message_kit = fields.MessageKit(dump_only=True)
+    signature = fields.Key(dump_only=True) # maybe we need a signature field?
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/enrico.py]: I would think that the Signature is different from a Key...

File: nucypher/characters/control/specifications/bob.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    # but bob.join_policy outputs {'policy_encrypting_key': 'OK'}
+
+
+class Retrieve(BaseSchema):
+    label = fields.Label(required=True, load_only=True)
+    policy_encrypting_key = fields.Key(required=True, load_only=True)
+    alice_verifying_key = fields.Key(required=True, load_only=True, )
+    message_kit = fields.MessageKit(required=True, load_only=True)
+
+    cleartexts = fields.List(fields.Str(), dump_only=True)
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/bob.py]: This output is base64 though, right?

File: nucypher/characters/control/specifications/fields/messagekit.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+from marshmallow import fields
+from base64 import b64decode, b64encode
+
+
+class MessageKitField(fields.Field):
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/fields/messagekit.py]: I guess this is really `UmbralMessageKit` unless you intend to use the same field type for `RevocationKit`...?

File: nucypher/characters/control/interfaces.py [hunk 7]
```diff
@@ -185,7 +190,7 @@ def join_policy(self, label: bytes, alice_verifying_key: bytes):
         Character control endpoint for joining a policy on the network.
         """
         self.bob.join_policy(label=label, alice_verifying_key=alice_verifying_key)
-        response = dict()  # {'policy_encrypting_key': ''}  # FIXME
+        response = {'policy_encrypting_key': 'OK'}  # FIXME
```
[the remark this answers is not in the document]
Comment (vepkenez) [reply nucypher/characters/control/interfaces.py]: yeah, I didn't do that, just moved it.  I think it's because `bob.join_policy` doesn't output anything useful and maybe that's a big project.

[the remark this answers is not in the document]
Comment (vepkenez) [reply nucypher/characters/control/interfaces.py]: Oh, and I just maintained it because the tests are testing for it.  I'm not changing the tests(yet).

File: nucypher/characters/control/controllers.py [hunk 8]
```diff
@@ -37,122 +38,93 @@ class CharacterControllerBase(ABC):
...
         self.emitter = self._emitter_class()
-        self.emitter.transport_serializer = self.serializer
 
         # Disables request & response serialization
         self.serialize = serialize
 
+    def get_serializer(self, interface_name: str):
+        return self.specification.get_serializer(inteface_name)
+
+    def _perform_action(self, action: str, request: dict) -> dict:
```
Comment (mswilkison) [inline nucypher/characters/control/controllers.py]: 👏

File: nucypher/characters/control/controllers.py [hunk 9]
```diff
@@ -37,122 +38,93 @@ class CharacterControllerBase(ABC):
...
     @character_control_interface
     def decrypt(self, request: dict):
-        result = super().decrypt(**self.serializer.load_decrypt_input(request=request))
-        response_data = self.serializer.dump_decrypt_output(response=result)
-        return response_data
+        return self._perform_action('decrypt', request)
 
     @character_control_interface
     def public_keys(self, request):
-        """
```
Comment (mswilkison) [inline nucypher/characters/control/controllers.py]: Should we keep the docstrings? (same question for other occurences)

File: nucypher/characters/control/specifications/exceptions.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class SpecificationError(ValueError):
+        """The protocol request is completely unusable"""
```
Comment (mswilkison) [inline nucypher/characters/control/specifications/exceptions.py]:
```suggestion
    """The protocol request is completely unusable"""
```

File: nucypher/characters/control/specifications/exceptions.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+class MissingField(SpecificationError):
+    """The protocol request can be deserialized by is missing required fields"""
+
+class InvalidInputField(SpecificationError):
+    """Response data does not match the output specification"""
```
Comment (mswilkison) [inline nucypher/characters/control/specifications/exceptions.py]:
```suggestion
    """Response data does not match the input specification"""
```

File: nucypher/characters/control/specifications/exceptions.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+class SpecificationError(ValueError):
+        """The protocol request is completely unusable"""
+
+class MissingField(SpecificationError):
+    """The protocol request can be deserialized by is missing required fields"""
```
Comment (mswilkison) [inline nucypher/characters/control/specifications/exceptions.py]:
```suggestion
    """The protocol request cannot be deserialized because it is missing required fields"""
```

File: nucypher/characters/control/specifications/fields/label.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    def _serialize(self, value, attr, obj, **kwargs):
+        return value.decode('utf-8')
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        if isinstance(value, bytes):
+            return value
+        return value.encode()
+
+    def _validate(self, value):
+        return True
```
Comment (mswilkison) [inline nucypher/characters/control/specifications/fields/label.py]: Is some validation logic supposed to happen here?

File: nucypher/characters/control/specifications/fields/messagekit.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+from typing import Union
```
Comment (mswilkison) [inline nucypher/characters/control/specifications/fields/messagekit.py]: Unused import

```suggestion
```

File: tests/characters/test_specifications.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import pytest
+from nucypher.characters.control.specifications import AliceSpecification
+from nucypher.characters.control.specifications.exceptions import SpecificationError
+
+
+def test_alice_spec_validate(federated_alice, federated_bob):
```
Comment (mswilkison) [inline tests/characters/test_specifications.py]: Test other character specs too?

File: nucypher/characters/control/specifications/fields/datetime.py [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+from nucypher.characters.control.specifications.fields.base import BaseField
+
+
+class DateTime(BaseField, fields.Field):
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return value.iso8601()
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        return maya.MayaDT.from_iso8601(iso8601_string=value)
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/fields/datetime.py]: I'm unsure, but would a validate function here help? i.e. catch invalid input earlier or is it just redundant since you would use Maya for the validation anyway? What's the code path for `_validate`?

File: nucypher/characters/control/specifications/bob.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    # but bob.join_policy outputs {'policy_encrypting_key': 'OK'}
+
+
+class Retrieve(BaseSchema):
+    label = fields.Label(required=True, load_only=True)
+    policy_encrypting_key = fields.Key(required=True, load_only=True)
+    alice_verifying_key = fields.Key(required=True, load_only=True, )
+    message_kit = fields.UmbralMessageKit(required=True, load_only=True)
+
+    cleartexts = fields.List(fields.Cleartext(), dump_only=True)
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/bob.py]: 👍

File: nucypher/characters/control/interfaces.py [hunk 18]
```diff
@@ -103,6 +107,7 @@ def grant(self,
               n: int,
               expiration: maya.MayaDT,
               value: int = None,
+              first_period_reward: int = None,
```
[the remark this answers is not in the document]
Comment (michwill) [reply nucypher/characters/control/interfaces.py]: Yes, I think this parameter got removed

File: nucypher/characters/control/specifications/fields/datetime.py (hunk 16 above)
Comment (vepkenez) [reply nucypher/characters/control/specifications/fields/datetime.py]: Yeah, the _validate runs on the  _output_ Of the _deserialize anyway so it would have failed.  It would be more to check that the date itself met some characteristic.

File: nucypher/characters/control/specifications/fields/treasuremap.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+class TreasureMap(BaseField, fields.Field):
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return b64encode(bytes(value)).decode()
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        return b64decode(value)
+
+    def _validate(self, value):
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/fields/treasuremap.py]: Can validate here simply be (?)

```python
try:
    treasure_map = TreasureMap.from_bytes(value):
    return True
except:
    return False
```

I'm unsure whether the `verify` kwarg for `from_bytes` can remain as `True` (default) or needs to be set to `False`.

Comment (vepkenez) [reply nucypher/characters/control/specifications/fields/treasuremap.py]: Yes, I'd love it if that worked but unfortunately importing `nucypher.policy.collections.Treasuremap` causes a circular import since Bob is imported into that same file... which I think should maybe be a TODO but maybe there is some reason I'm not aware of.

**Comment by vepkenez** [comment]
> Unsure how I feel about the CLI integration - a bit of indirection - need to think a bit more about it.

I agree @derekpierre.  It's not ideal.  It was better before the cleanup-options PR.  I think in the long run, the path is going to be to remove some of the complexity and have all of the CLI options be specified through similar schemas so it will actually be very simple.  

so instead of 
```
alice_verifying_key = fields.Key(
         load_only=True, required=True,
         click=click.option(
             '--alice-verifying-key',
             help="Alice's verifying key as a hexadecimal string",
             required=True, type=click.STRING,))
```

we'll have
```
alice_verifying_key = fields.Key(load_only=True, required=True, help="Alice's verifying key as a hexadecimal string")
```
And then the click stuff would be generated from that. 
Because of the recent changes that happened concurrently to the work in this PR, I just folded it in in a way that would provide exactly matching results.

@KPrasch and I are planning to expand this concept to move most of the ad-hoc code that lives in the cli files into interfaces that can then be accessed through any entry point... but later.

**Comment by KPrasch** [comment]
Rebased over `1234006780ad00952629b5825398625a0428319e`

File: nucypher/characters/control/interfaces.py [hunk 20]
```diff
@@ -1,59 +1,31 @@
...
-        # Validate response
-        instance.validate_response(response=response, interface_name=interface_name)
+    def callable(func):
+        func._schema = schema()
 
-        # Record duration
-        responding = maya.now()
-        duration = responding - received
+        @functools.wraps(func)
+        def wrapped(*args, **kwargs):
```
Comment (KPrasch) [inline nucypher/characters/control/interfaces.py]: This inner wrapper does not do anything?

File: nucypher/characters/control/interfaces.py [hunk 21]
```diff
@@ -62,13 +34,27 @@ def __init__(self, character=None, *args, **kwargs):
         self.character = character
         super().__init__(*args, **kwargs)
 
+    @classmethod
+    def connect_cli(cls, action):
```
Comment (KPrasch) [inline nucypher/characters/control/interfaces.py]: A comment here in is sorely needed.

File: nucypher/characters/control/controllers.py [hunk 22]
```diff
@@ -178,27 +71,60 @@ def __init__(self,
         # Internals
         self._transport = None
 
-        # Hard-wire the character's output flow to the Emitter
-        self._internal_controller = character_controller
-        self._internal_controller.emitter = self.emitter
+        self.interface = interface
+
+        def set_method(name):
```
Comment (KPrasch) [inline nucypher/characters/control/controllers.py]: Needs follow up for deeper integration IMO - perhaps move to a method explaining the polymorphic approach and implicit routing implicitly occurring here.   I think this can be named `self.bind_interface()`

File: nucypher/characters/control/controllers.py [hunk 23]
```diff
@@ -178,27 +71,60 @@ def __init__(self,
...
+
+            def wrapper(request=None, **kwargs):
+                request = request or kwargs
```
Comment (KPrasch) [inline nucypher/characters/control/controllers.py]: This is a shim for optionally passing kwargs to a controller rather than a dictionary (unsupported behaviors), which needs to be more formally enforced.

File: nucypher/characters/control/controllers.py [hunk 24]
```diff
@@ -178,27 +71,60 @@ def __init__(self,
...
 
-        super().__init__(control_serializer=self._internal_controller.serializer)
+        for method_name in self._get_interfaces().keys():
+            set_method(method_name)
+
+        super().__init__(*args, **kwargs)
 
         self.log = Logger(app_name)
 
+    def _get_interfaces(self):
```
Comment (KPrasch) [inline nucypher/characters/control/controllers.py]: Let's embed this method on a base class of `Interfaces`

File: nucypher/characters/control/controllers.py [hunk 25]
```diff
@@ -178,27 +71,60 @@ def __init__(self,
...
+        return {
+            name: method for name, method in
+            inspect.getmembers(
+                self.interface,
+                predicate=inspect.ismethod)
+            if hasattr(method, '_schema')
```
Comment (KPrasch) [inline nucypher/characters/control/controllers.py]: Perhaps move the suffix to somewhere more conspicuous like a class level variable `Interface.PREDICATE`.

File: nucypher/characters/control/controllers.py [hunk 26]
```diff
@@ -178,27 +71,60 @@ def __init__(self,
...
     def test_client(self):
         return NotImplemented
 
 
+class CLIController(CharacterControlServer):
+
+    _emitter_class = StdoutEmitter
+
+    def make_control_transport(self):
+        return
```
Comment (KPrasch) [inline nucypher/characters/control/controllers.py]: Seems awkward to disable this method, perhaps indicating additional refactoring needed to fully realize the critique of this PR.

File: nucypher/cli/options.py [hunk 27]
```diff
@@ -53,6 +51,9 @@
 option_teacher_uri = click.option('--teacher', 'teacher_uri', help="An Ursula URI to start learning from (seednode)", type=click.STRING)
 _option_middleware = click.option('-Z', '--mock-networking', help="Use in-memory transport instead of networking", count=True)
 
+# Avoid circular input
```
Comment (KPrasch) [inline nucypher/cli/options.py]: This is not a common option - So this is just a semi-temporary location for this flag until the full refactor can be performed.

File: tests/characters/control/blockchain/test_rpc_control_blockchain.py [hunk 28]
```diff
@@ -56,6 +63,14 @@ def test_alice_rpc_character_control_create_policy(alice_rpc_test_client, create
         assert rpc_response.success is True
         assert rpc_response.id == response_id
 
+def test_alice_rpc_character_control_bad_input(alice_rpc_test_client, create_policy_control_request):
```
Comment (derekpierre) [inline tests/characters/control/blockchain/test_rpc_control_blockchain.py]: do we only test bad input for alice? what about others?

File: nucypher/characters/control/specifications/alice.py [hunk 29]
```diff
@@ -36,9 +39,26 @@ class PolicyBaseSchema(BaseSchema):
...
+    )
+
     # output
     policy_encrypting_key = fields.Key(dump_only=True)
 
+    @validates_schema
+    def check_n_gte_m(self, data, **kwargs):
+        # ensure that n is greater than or equal to m
+        if not data['n'] >= data['m']:
+            raise InvalidArgumentCombo
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/alice.py]: We should be descriptive here i.e. include the reason text.

File: nucypher/characters/control/specifications/alice.py [hunk 30]
```diff
@@ -36,9 +39,26 @@ class PolicyBaseSchema(BaseSchema):
...
+    rate = fields.Wei(
+        load_only=True,
+        required=False,
+        click=options.option_rate
+    )
+
     # output
     policy_encrypting_key = fields.Key(dump_only=True)
 
+    @validates_schema
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/alice.py]: 👍 - it's cool that all of the validation logic does not have to be done in one function.

File: tests/characters/test_specifications.py [hunk 31]
```diff
@@ -1,10 +1,50 @@
...
+    result = GrantPolicy().load(data)
+    assert result['label'] == b'cats the animal'
+
+    # validate that m > n fails validation
+    data['m'] = data['n'] + 19
     with pytest.raises(SpecificationError) as e:
-        _result = CreatePolicy().load(dict())
+        GrantPolicy().load(data)
+
+    breakpoint()
```
Comment (derekpierre) [inline tests/characters/test_specifications.py]:
```suggestion
```

Comment (vepkenez) [reply tests/characters/test_specifications.py]: heh. yeah :|

File: tests/characters/control/blockchain/test_rpc_control_blockchain.py (hunk 28 above)
Comment (vepkenez) [reply tests/characters/control/blockchain/test_rpc_control_blockchain.py]: It's not really in this PR's scope for me to write tests for every single character's input.  I'm just testing the implementation.  The code paths are the same for all the other characters so this test covers all of them to some degree.

File: nucypher/characters/lawful.py [hunk 32]
```diff
@@ -384,64 +374,55 @@ def public_keys():
...
 
         @alice_flask_control.route('/derive_policy_encrypting_key/<label>', methods=['POST'])
         def derive_policy_encrypting_key(label) -> Response:
             """
             Character control endpoint for deriving a policy encrypting given a unicode label.
             """
-            response = controller(interface=controller._internal_controller.derive_policy_encrypting_key,
-                                  control_request=request,
-                                  label=label)
+            response = controller(method_name='derive_policy_encrypting_key', control_request=request, label=label)
```
[the remark this answers is not in the document]
Comment (vepkenez) [reply nucypher/characters/lawful.py]: This is some vestigial thing left over from the web controller maybe...  I'm just keeping it as is.

File: tests/characters/control/blockchain/test_rpc_control_blockchain.py (hunk 28 above)
Comment (derekpierre) [reply tests/characters/control/blockchain/test_rpc_control_blockchain.py]: It feels like our list of follow-ups to this PR is growing. If not done already, perhaps creating and linking follow-up issue(s) would be best.
1. CLI parameter thing we talked about during the walkthrough
2. Expanded character tests, or modifying the tests to be generic/repeatable for other character specifications.

**Comment by derekpierre** [comment]
@KPrasch , @vepkenez : The following comment is hindering me from approving this PR: https://github.com/nucypher/nucypher/pull/1555/files#r363467150 - any thoughts there? Perhaps I'm mistaken? Once resolved I'll approve.

File: nucypher/characters/base.py [hunk 33]
```diff
@@ -116,7 +116,9 @@ def __init__(self,
 
         #
         # Operating Mode
-
+        if hasattr(self, '_interface_class'):  # TODO: have argument about meaning of 'lawful'
+            #                                         and whether maybe only Lawful characters have an interface
+            self.interface = self._interface_class(character=self)
```
Comment (KPrasch) [inline nucypher/characters/base.py]: This is only needed when using a controller, consider moving this out of init to prevent interface creation for characters that use just the plain python API.

File: nucypher/characters/control/interfaces.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/control/interfaces.py]: #1459 is already merged into matster, so if this PR is currently based over the latest, and this variable is unused within this function it can be safely removed.

File: nucypher/characters/control/interfaces.py [hunk 34]
```diff
@@ -171,29 +162,28 @@ def decrypt(self, label: bytes, message_kit: bytes) -> dict:
...
 
+    @attach_schema(bob.JoinPolicy)
     def join_policy(self, label: bytes, alice_verifying_key: bytes):
         """
         Character control endpoint for joining a policy on the network.
         """
-        self.bob.join_policy(label=label, alice_verifying_key=alice_verifying_key)
-        response = dict()  # {'policy_encrypting_key': ''}  # FIXME
+        self.character.join_policy(label=label, alice_verifying_key=alice_verifying_key)
+        response = {'policy_encrypting_key': 'OK'}  # FIXME
```
Comment (KPrasch) [inline nucypher/characters/control/interfaces.py]: Do we have the PEK from the result of the above call?

File: nucypher/characters/control/specifications/alice.py (hunk 30 above)
Comment (KPrasch) [reply nucypher/characters/control/specifications/alice.py]: Continued scoping: Handle errors and error output via `Emitters`

File: nucypher/characters/control/specifications/exceptions.py [hunk 35]
```diff
@@ -0,0 +1,14 @@
...
+class MissingField(SpecificationError):
+    """The protocol request cannot be deserialized because it is missing required fields"""
+
+class InvalidInputField(SpecificationError):
+    """Response data does not match the input specification"""
+
+class InvalidOutputField(SpecificationError):
+    """Response data does not match the output specification"""
+
+class InvalidArgumentCombo(SpecificationError):
```
Comment (jMyles) [inline nucypher/characters/control/specifications/exceptions.py]: This is interesting.  We've often discussed how to formalize "xor" arguments - we have different mutually exclusivity logic in different parts of the codebase.  Perhaps this is the beginning of something normal.

File: nucypher/characters/base.py [hunk 36]
```diff
@@ -515,11 +517,20 @@ def _set_checksum_address(self):
 
         self._checksum_address = public_address
 
-    def make_rpc_controller(drone, crash_on_error: bool = False):
-        app_name = bytes(drone.stamp).hex()[:6]
+    def make_rpc_controller(self, crash_on_error: bool = False):
```
[the remark this answers is not in the document]
Comment (tuxxy) [reply nucypher/characters/base.py]: This. I really enjoyed reading `drone` as it made a lot of sense when reading.

File: nucypher/characters/control/emitters.py [hunk 37]
```diff
@@ -232,10 +237,8 @@ def exception(drone_character,
             raise e
         return drone_character.sink(str(e), status=response_code)
 
-    def ipc(drone_character, response, request_id, duration) -> Response:
-        assembled_response = drone_character.assemble_response(response=response,
-                                                               request_id=request_id,
-                                                               duration=duration)
+    def respond(drone_character, response) -> Response:
```
Comment (tuxxy) [inline nucypher/characters/control/emitters.py]: Solid rename. :+1:

File: nucypher/characters/lawful.py [hunk 38]
```diff
@@ -136,7 +136,7 @@ def __init__(self,
                                             checksum_address=checksum_address)
 
         if is_me and controller:
-            self.controller = self._controller_class(alice=self)
+            self.make_cli_controller()
```
Comment (tuxxy) [inline nucypher/characters/lawful.py]: I really like how this evolved. :+1:

File: nucypher/characters/control/specifications/alice.py [hunk 39]
```diff
@@ -0,0 +1,121 @@
...
+
+    @validates_schema
+    def check_n_gte_m(self, data, **kwargs):
+        # ensure that n is greater than or equal to m
+        if not data['n'] >= data['m']:
+            raise InvalidArgumentCombo("N must be greater than or equal to M")
+
+    @validates_schema
+    def check_rate_or_value_not_both(self, data, **kwargs):
+        if data.get('rate') and data.get('value'):
```
[the remark this answers is not in the document]
Comment (vepkenez) [reply nucypher/characters/control/specifications/alice.py]: hmm maybe.  I'll make that change!

File: tests/characters/test_specifications.py [hunk 40]
```diff
@@ -0,0 +1,48 @@
...
+
+    # remove value and now it works
+    del data['value']
+    result = GrantPolicy().load(data)
+    assert result['label'] == b'cats the animal'
+
+    # validate that m > n fails validation
+    data['m'] = data['n'] + 19
+    with pytest.raises(SpecificationError) as e:
+        GrantPolicy().load(data)
```
[the remark this answers is not in the document]
Comment (vepkenez) [reply tests/characters/test_specifications.py]: yes this test is testing the specification validation, not the granting.

File: nucypher/characters/control/specifications/fields/label.py (hunk 13 above)
Comment (cygnusv) [reply nucypher/characters/control/specifications/fields/label.py]: > Is some validation logic supposed to happen here?

Do we want to accept empty labels (`""`)? Are there other encodings that are outside UTF8?

File: nucypher/characters/control/specifications/fields/misc.py [hunk 41]
```diff
@@ -0,0 +1,34 @@
...
+
+class String(BaseField, fields.String):
+    pass
+
+
+class List(BaseField, fields.List):
+    pass
+
+
+class Integer(BaseField, fields.Integer):
```
Comment (cygnusv) [inline nucypher/characters/control/specifications/fields/misc.py]: I'd create a PositiveInteger subclass with a `_validate()`, as most (all?) integer fields are actually intended to be positive.

File: nucypher/characters/control/specifications/fields/treasuremap.py [hunk 42]
```diff
@@ -0,0 +1,30 @@
...
+        # TODO: we don't currently accept any treasuremap as input so this is not
+        # used.
```
Comment (cygnusv) [inline nucypher/characters/control/specifications/fields/treasuremap.py]: This is going to change in birdcage. Should this PR be prepared for that? Unless @jmyles has already done this there.

File: nucypher/cli/commands/alice.py [hunk 43]
```diff
@@ -393,7 +384,7 @@ def grant(general_config,
         if any((value, rate)):
             raise click.BadOptionUsage(option_name="--value, --rate",
                                        message="Can't use --value or --rate with a federated Alice.")
-    elif not (bool(value) ^ bool(rate)):
+    if bool(value) and bool(rate):
```
Comment (cygnusv) [inline nucypher/cli/commands/alice.py]: Why this change? Now it misses when `bool(value)==False` and `bool(rate)==False`

**Comment by cygnusv** [comment]
Also, I think the commit history in this PR could be cleaner. For example, there are several commits that can be easily squashed.

File: nucypher/characters/control/specifications/fields/misc.py (hunk 41 above)
Comment (vepkenez) [reply nucypher/characters/control/specifications/fields/misc.py]: already done.  next push!

**Comment by derekpierre** [comment]
I agree with @cygnusv that some commits can probably be squashed for a cleaner history.
